language/diem-vm/src/parallel_executor/dependency_analyzer.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    data_cache::StateViewCache,
+    diem_transaction_executor::{preprocess_transaction, PreprocessedTransaction},
+    logging::AdapterLogSchema,
+    DiemVM,
+};
+use diem_state_view::StateView;
+use diem_types::{
+    access_path::AccessPath,
+    account_address::AccountAddress,
+    transaction::{Transaction, TransactionArgument, TransactionPayload},
+};
+use move_core_types::vm_status::VMStatus;
+use std::collections::{HashMap, HashSet};
+
+// Structure that holds infered read/write sets
+
+#[derive(Debug, Clone)]
+enum ScriptReadWriteSetVar {
+    Const,
+    Param(usize),
+}
+
+#[derive(Debug)]
+struct ScriptReadWriteSet {
```

Comment:
To avoid duplicate definitions/boilerplate conversion code, I think it's probably better to let the analysis define this data structure and ask it to implement a higher-level API like

```
trait ReadWriteSetAnalysis [
/// Substitute concrete values from a transaction + the global state into `self` to produce a set of concrete access paths
fn sub(&self, tx: RawTransaction, f: AccessPath -> Option<Value>) -> Set<(AccessPath, AccessKind)>

/// Return true if `self` may write an access path that is read by `other`
fn may_have_write_read_conflict(&self, other: &Self) -> bool
```

Reply:
Lemme create a draft version of the trait and put it in a fixup commit and see if you like it!